proposals/poc/witx/proposal_signatures.witx

```diff
@@ -0,0 +1,303 @@
+;; WASI Cryptography API
+
+(typename $errno
+  (enum u16
+    $success
+    $notavailable
+    $invalidkey
+    $verificationfailed
+    $rngerror
+    $algorithmfailure
+    $invalidsignature
+    $closed
+    $invalidhandle
+    $overflow
+  )
+)
+
+(typename $keypair_encoding
+  (enum u16
+    $raw
+    $pkcs8
+    $der
+    $pem
+  )
+)
+
+(typename $publickey_encoding
+  (enum u16
+    $raw
+    $hex
+    $base64_original
+    $base64_original_nopadding
+    $base64_urlsafe
+    $base64_urlsafe_nopadding
+  )
+)
+
+(typename $signature_encoding
+  (enum u16
+    $raw
+    $hex
+    $base64_original
+    $base64_original_nopadding
+    $base64_urlsafe
+    $base64_urlsafe_nopadding
+    $der
+  )
+)
+
+(typename $version
+  (int u64
+;;; Key doesn't support versioning
+    (const $unspecified 0xff00000000000000)
+;;; Retrieve the latest version of a key
+    (const $latest 0xff00000000000001)
+;;; Perform an operation over all versions of a key
+    (const $all 0xff00000000000002)
+  )
+)
+
+(typename $size (@witx usize))
+
+(typename $array_output (handle))
+(typename $signature_op (handle))
+(typename $signature_keypair_builder (handle))
+(typename $signature_keypair (handle))
+(typename $signature_state (handle))
+(typename $signature (handle))
+(typename $signature_publickey (handle))
+(typename $signature_verification_state (handle))
+
+(module $wasi_ephemeral_crypto
+  (import "memory" (memory))
+
+;;; Return the length of an array_output object
+  (@interface func (export "array_output_len")
+    (param $array_output $array_output)
+    (result $len $size)
+  )
+
+;;; Copy an array_output into an application-allocated buffer
+;;; The array_output handle becomes invalid after this operation
+  (@interface func (export "array_output_pull")
+    (param $array_output $array_output)
+    (param $buf (@witx pointer u8))
+    (param $buf_len $size)
+    (result $error $errno)
+  )
+
+;;; Create a context for a signature-related operation
+  (@interface func (export "signature_op_open")
+    (param $op_str string)
+    (result $error $errno)
+    (result $signature_op $signature_op)
+  )
+
+;;; Destroy the signature context
+  (@interface func (export "signature_op_close")
+    (param $signature_op $signature_op)
+    (result $error $errno)
+  )
+
+;;; Create a context to construct a key pair
+  (@interface func (export "signature_keypair_builder_open")
+    (param $signature_op $signature_op)
+    (result $error $errno)
+    (result $handle $signature_keypair_builder)
+  )
+
+;;; Destroy a key pair builder
+  (@interface func (export "signature_keypair_builder_close")
+    (param $kp_builder $signature_keypair_builder)
+    (result $error $errno)
+  )
+
+;;; Generate a new key pair
+;;; This function may return errno.notavailable if key generation
+;;; is not support by the host for the chose algorithm
```

Comment:
"is not supported by the host for the chosen algorithm"?

Reply:
Good catch, fixed, thanks!